### PR TITLE
Added QuerystringPreprocessor

### PR DESCRIPTION
**Pull request by jfrommann**
Please have a look at the StandardQuerystringPreprocessor.
- Should we leave german replacements in the default replacements?
- Should it be the default preprocessor bean in DefaultQueryConfiguration?

### CODE REVIEW

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerAutoBracket.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+package de.picturesafe.search.elasticsearch.connect.query.preprocessor;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+public class OptimizerAutoBracket implements QueryTokenOptimizer {
```
Comment (araithel) [inline src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerAutoBracket.java]: I would prefer 'AutoBracketOptimizer'

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerDefaultOperator.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * limitations under the License.
+ */
+package de.picturesafe.search.elasticsearch.connect.query.preprocessor;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class OptimizerDefaultOperator implements QueryTokenOptimizer {
```
Comment (araithel) [inline src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerDefaultOperator.java]: I would prefer 'DefaultOperatorOptimizer'

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/StandardQuerystringPreprocessor.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    static {
+        DEFAULT_REPLACEMENTS = new HashMap<>();
+        DEFAULT_REPLACEMENTS.put(",", " " + TOKEN_OR + " ");  // Blanks are necessary, do not remove!
+        DEFAULT_REPLACEMENTS.put("|", TOKEN_OR);
+        DEFAULT_REPLACEMENTS.put("oder", TOKEN_OR);
+        DEFAULT_REPLACEMENTS.put("or", TOKEN_OR);
+        DEFAULT_REPLACEMENTS.put("&", TOKEN_AND);
+        DEFAULT_REPLACEMENTS.put("+", TOKEN_AND);
+        DEFAULT_REPLACEMENTS.put("und", TOKEN_AND);
+        DEFAULT_REPLACEMENTS.put("and", TOKEN_AND);
```
Comment (araithel) [inline src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/StandardQuerystringPreprocessor.java]: I would provide a configuration option (also to support more languages), for example in elasticsearch.properties something like:
standard_querystring_preprocessor.replacement_tokens.and=& + and und

File: src/main/java/de/picturesafe/search/spring/configuration/DefaultQueryConfiguration.java [hunk 4]
```diff
@@ -52,6 +54,11 @@ public QueryConfiguration queryConfiguration() {
         return new QueryConfiguration();
     }
 
+    @Bean
+    public QuerystringPreprocessor querystringPreprocessor() {
+        return new NoOpQuerystringPreprocessor();
```
Comment (araithel) [inline src/main/java/de/picturesafe/search/spring/configuration/DefaultQueryConfiguration.java]: I would configure the StandardQueryStringPreprocessor and provide a configuration option, something like 'querystring_preprocessor.enabled=false'

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerAutoBracket.java (hunk 1 above)
Comment (jfrommann) [reply src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerAutoBracket.java]: done

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerDefaultOperator.java (hunk 2 above)
Comment (jfrommann) [reply src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/OptimizerDefaultOperator.java]: done

File: src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/StandardQuerystringPreprocessor.java (hunk 3 above)
Comment (jfrommann) [reply src/main/java/de/picturesafe/search/elasticsearch/connect/query/preprocessor/StandardQuerystringPreprocessor.java]: done

File: src/main/java/de/picturesafe/search/spring/configuration/DefaultQueryConfiguration.java (hunk 4 above)
Comment (jfrommann) [reply src/main/java/de/picturesafe/search/spring/configuration/DefaultQueryConfiguration.java]: done

**Pull request merged**